examples/blocked_cholesky_multigpu.py

```diff
@@ -0,0 +1,265 @@
+  This is a less naive version of dpotrf with one level of blocking.
+  Blocks are currently assumed to evenly divide the axes lengths.
+  The input array 4 dimensional. The first and second index select
+  the block (row first, then column). The third and fourth index
+  select the entry within the given block.
+  """
+  if a.shape[0] * a.shape[2] != a.shape[1] * a.shape[3]:
+    raise ValueError("A square matrix is required.")
+  if a.shape[0] != a.shape[1]:
+    raise ValueError("Non-square blocks are not supported.")
+
+  # Define task spaces
+  gemm1 = TaskSpace("gemm1")        # Inter-block GEMM
+  subcholesky = TaskSpace("subcholesky")  # Cholesky on block
+  gemm2 = TaskSpace("gemm2")        # Inter-block GEMM
+  solve = TaskSpace("solve")        # Triangular solve
+
+  for j in range(a.shape[0]):
+    for k in range(j):
+      # Inter - block GEMM
+      @spawn(gemm1[j, k], [solve[j, k]], placement=[gpu(j%num_gpus)])
+      def t1():
+        out = get_gpu_memory(j, j, num_gpus)
+        rhs = get_gpu_memory(j, k, num_gpus)
+        out = update(rhs, rhs, out)
+        set_gpu_memory_from_gpu(j, j, num_gpus, out)
+
+    # Cholesky on block
+    @spawn(subcholesky[j], [gemm1[j, 0:j]], placement=[gpu(j%num_gpus)])
+    def t2():
+      dblock = get_gpu_memory(j, j, num_gpus) 
+      dblock = cholesky(dblock)
+      set_gpu_memory_from_gpu(j, j, num_gpus, dblock)
+
+    for i in range(j+1, a.shape[0]):
+      for k in range(j):
+        # Inter - block GEMM
+        @spawn(gemm2[i, j, k], [solve[j, k], solve[i, k]], placement=[gpu(i%num_gpus)])
+        def t3():
+          out = get_gpu_memory(i, j, num_gpus)
+          rhs1 = get_gpu_memory(i, k, num_gpus)
+          rhs2 = get_gpu_memory(j, k, num_gpus)
+          out = update(rhs1, rhs2, out)
+          set_gpu_memory_from_gpu(i, j, num_gpus, out)
+
+      # Triangular solve
+      @spawn(solve[i, j], [gemm2[i, j, 0:j], subcholesky[j]], placement=[gpu(i%num_gpus)])
+      def t4():
+        factor = get_gpu_memory(j, j, num_gpus)
+        panel  = get_gpu_memory(i, j, num_gpus)
+        out = ltriang_solve(factor, panel)
+        set_gpu_memory_from_gpu(i, j, num_gpus, out)
+
+  return subcholesky[a.shape[0]-1]
+
+def allocate_gpu_memory(i:int, r:int, n:int, b:int):
+  with cp.cuda.Device(i):
+    logger.debug("\tAllocate device:", i, "...")
+    prealloced = cp.ndarray([r, n // b, b, b])
+    gpu_arrs.append(prealloced)
+
+def get_gpu_memory(i:int, j:int, num_gpus:int):
+  dev_id   = i % num_gpus
+  local_id = i // num_gpus
+  src = gpu_arrs[dev_id][local_id][j]
+  dst = clone_here(src)
+  return dst
+
+def set_gpu_memory_from_gpu(i:int, j:int, num_gpus:int, v):
+  dev_id   = i % num_gpus
+  local_id = i // num_gpus
+  gpu_arrs[dev_id][local_id][j] = v
+
+def set_gpu_memory_from_cpu(a, num_gpus):
+  for j in range(a.shape[0]):
+    dev_id   = j % num_gpus 
+    local_id = j // num_gpus 
+    with cp.cuda.Device(dev_id):
+      gpu_arrs[dev_id][local_id] = cp.array(a[j], copy=True)
```

Comment:
These functions can be defined in a class along with `cholesky_blocked_inplace` so the class can capture the value of `num_gpu` and clean up this code a bit.

Also I think the mess here means that a multi-device array abstraction type would be useful. This simple approach may well not be general enough, but we clearly need a way to work with and directly index arrays that are actually multiple arrays on different devices. User code should never have to have junk like this.

Reply:
Yah, we need to think about how to do this better. A multi-device array abstraction is a huge code cost, so we might be better served by working out some kind of hook into dask to leverage what they've already implemented in some way. Alternatively we could work on expanding our partitioning functionality. IDK. It's worth thinking about.